### ISSUE TITLE
Use of atoms for index names leads to exhaustion

### ISSUE THREAD

**Issue by JonathonMA** [rabbitmq/rabbitmq-management#245]
We've experienced several cases where BEAM crashes with:

```
no more index entries in atom_tab (max=1048576)
```

Examining the crashlog I can see that the majority of the atom table matches these two patterns:

```
507960 #_key_index
507960 #_index
```

I did some investigation and since atoms aren't garbage collected they shouldn't be dynamically generated. Looks like these patterns are generated and converted to atoms in the links below, so I'm guessing these are the culprits.

https://github.com/rabbitmq/rabbitmq-management/blob/f2b3ebddadc35cca671eafc652cff7cdeac01c99/src/rabbit_mgmt_stats_tables.erl#L219-L220

https://github.com/rabbitmq/rabbitmq-management/blob/f2b3ebddadc35cca671eafc652cff7cdeac01c99/src/rabbit_mgmt_stats_tables.erl#L276-L277

**Comment by michaelklishin** [comment]
@dcorbacho how realistic is it to change index key naming schemes before the planned schema changes in #236?

**Comment by dcorbacho** [comment]
@michaelklishin This only affects the temporary tables created to aggregate data during the querying, otherwise our stat tables are fixed to 26 x (table, table_index, table_key_index).

I will check if we can use unnamed tables for the aggregation, passing the table references only for these.

**Comment by clain23** [comment]
We have exactly the same crashes on 3.6.2.

**Comment by michaelklishin** [comment]
#246 will be in `3.6.4 M1` later this week.

**Comment by michaelklishin** [comment]
As a workaround it's possible to bump the atom table size with the `+t` VM flag (default is `1048576`).

**Comment by jcomellas** [comment]
@michaelklishin We have experienced the same problem in production with RabbitMQ 3.6.5. I have the `erl_crash.dump` file from when this happened. Would it be useful if we sent it to you?

**Comment by michaelklishin** [comment]
@jcomellas we are working on a new management plugin which stores the stats very differently, so probably not. Thank you.

**Comment by jcomellas** [comment]
@michaelklishin Do you have an ETA for this? Is the recommended solution for the time being to just bump the size of the atom table?

**Comment by michaelklishin** [comment]
Yes, just bump the table size.

We hope it will ship in `3.6.7` or `3.6.8`.

**Comment by dieswaytoofast** [comment]
Good to know - tx @michaelklishin

**Comment by uvzubovs** [comment]
This issue is marked as fixed, but Rabbit 3.6.5 still runs out of memory with the same symptom as 3.6.3 when deployed to PCF, and we were told it was due to atom table exhaustion. We even put in a config param to increase atom table size, which helped. What's unique about PCF, is that it runs a metrics collector task that periodically hits Rabbit management API.

**Comment by michaelklishin** [comment]
@uvzubovs multiple users no longer have this problem after moving to 3.6.5 or 3.6.6. Periodic polling of stats is a very common thing, PCF or not. Anyhow, #236 ships in 3.6.7 and we haven't seen atom table grow large with it either.
